File: 0239-sliding-window-maximum/0239-sliding-window-maximum.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        deque<int> d;
        vector<int> ans;
        for (int i=0; i<nums.size(); i++) {
            if (!d.empty() and d.front() == i-k) d.pop_front();
            while (!d.empty() and nums[d.back()] < nums[i]) d.pop_back();
            d.push_back(i);
            if (i >= k-1) ans.push_back(nums[d.front()]);
        }
        return ans;
    }
};

File: 0239-sliding-window-maximum/0239-sliding-window-maximum.py
from sortedcontainers import SortedList

class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        cur = SortedList(nums[:k])
        ans = [cur[-1]]
        for i in range(k, len(nums)):
            cur.remove(nums[i-k])
            cur.add(nums[i])
            ans.append(cur[-1])
        return ans